### ISSUE TITLE
Grid restrictions

### ISSUE THREAD

**Issue by andrewGhazi** [onnela-lab/gptools#25]
Hi there, I'd like to make use of this package on a spatial transcriptomics dataset and wanted to ask about whether it could work with my data. The two details I'm wondering about are:

1. My data lie on a *triangular* grid. Will `gp_inv_rfft` and friends work on any regularly spaced pattern or does it specifically require points that lie on the corners of tiled squares?
2. The data fall in an irregular shape that also include holes. Am I right to assume that I'll need to pad in those irregularities?

The plot of measured points below demonstrates both of these issues. I'd appreciate any help you could provide. I started to dig into the math because it feels like both of these problems should be solvable in theory, but I thought I'd ask since I imagine it will be immediately obvious to the authors if this could be made to work or not. Thanks!

![hex](https://github.com/user-attachments/assets/8d411e18-6eca-45ee-9d78-4134554b704e)

**Comment by tillahoffmann** [comment]
Hey, thanks for getting in touch.

1. The methods require a regularly spaced, rectangular grid, but a triangular grid can be accommodated by oversampling the space (I've attached an example below). This will of course "waste" some resources, but we've found it to run faster than a standard implementation because of the favorable scaling with sample size. What sort of grid sizes are you working with?
2. Assuming you use a regular grid as in the illustration below you won't need to explicitly take care of the holes. Say the Gaussian process is $z_{ij}$ at row $i$ and column $j$ in the grid. Then wherever you have data $y_{ij}$, you can include a sampling statement $y_{ij}\sim\mathsf{SomeLikelihood}\left(z_{ij}\right)$. If you don't have data for a location, all you need to do is omit the sampling statement.

Another option to consider might be a nearest neighbor Gaussian process (see [here](https://gptools-stan.readthedocs.io/en/latest/docs/background.html#sparse-approximation) for details). The Fourier methods are generally quite a bit faster though (see Figure 2 on page 13 [here](https://arxiv.org/pdf/2301.08836)).

I saw you work for HMS. I'm just across the road at HSPH. Happy to meet up if you're around on campus.

![386876321-8d411e18-6eca-45ee-9d78-4134554b704e](https://github.com/user-attachments/assets/9b73e3e4-adff-4808-8d75-949669c34b99)

**Comment by tillahoffmann** [comment]
> I take it then that rectangular (as opposed to perfectly square) grids are okay?

Yes, in practice there are two ways to deal with the non-square grid.

1. Use a vector of scale parameters (`vector<lower=0> [2] length_scales;` in the parameter block) to account for the different effective length scales and then pass the vector to the evaluation of the kernel, e.g., [`gp_periodic_exp_quad_cov_rfft2`](https://gptools-stan.readthedocs.io/en/latest/docs/interface/fft.html#7218a02a-9caf-4caf-8452-f368a60de55a). However, this approach allows for different length scales in the physical space because the parameters are independent.
2. Use a scalar length scale, construct a length scale vector based on the known aspect ratio of the grid, and then proceed as above. This ensures that the kernel is isotropic in the physical space. E.g., something like this.

```stan
parameters {
  real<lower=0> length_scale;
}

transformed parameters {
  vector<lower=0> [2] length_scales = (2 * length_scale, length_scale)';
}
```

The factor of two accounts for the oversampling in the horizontal direction. The grid is more densely spaced in this direction so, if we "rescaled" things to get a square grid, the correlation length has to be longer by a factor of two. 

I *think* that's right, but I always get myself confused about where the factor of two should go. One way to double check would be to sample from the GP prior and make sure the posterior samples look as expected.

> The number of spots in a sample can go up to 25k [...] My hope is to model all of these features plus some other complicating factors using joint correlated GPs, though I'm unsure if this is computationally feasible.

Yes, that scale may be tricky. How many correlated GPs are you using and what's the correlation structure?

If Stan isn't able to handle the scale, variational inference using [numpyro](https://github.com/pyro-ppl/numpyro) may be an alternative. The components for FFT GPs are there (see e.g., pyro-ppl/numpyro#1762), but there's a bit of legwork to be done. Stan also implements variational inference, but it's a bit brittle.

> I'll need a week or so to play around with this, but yes a meet up would be great. My office is in Countway so I think we're right next door. I'll send an email to your HSPH address to schedule.

Sounds great!

**Comment by tillahoffmann** [comment]
I'm closing this one for now, but please feel free to re-open if you want to discuss the grid restrictions further.
